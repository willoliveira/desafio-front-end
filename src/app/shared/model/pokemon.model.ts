import { IRessource } from "./ressource.model";

export enum EnumPokemonTypes {
	'normal',
	'fighting',
	'flying',
	'poison',
	'ground',
	'rock',
	'bug',
	'ghost',
	'steel',
	'fire',
	'water',
	'grass',
	'electric',
	'psychic',
	'ice',
	'dragon',
	'dark',
	'fairy',
	'unknown',
	'shadow'
}

export interface IPokemonAbilities {
	ability: IRessource;
	is_hidden: boolean;
	slot: number;
}

export interface IPokemonStats {
	base_stat: number;
	effort: number;
	stat: IRessource;
}

export interface IPokemonTypes {
	slot: number;
	type: IRessource;
}

export interface IPokemonSprites {
	back_default: string;
	back_female: string;
	back_shiny: string;
	back_shiny_female: string;
	front_default: string;
	front_female: string;
	front_shiny: string;
	front_shiny_female: string;
}

export interface IPokemon {
	id: number;
	base_experience: number;
	name: string;
	sprites: IPokemonSprites;
	weight: number;
	height: number;
	types: IPokemonTypes[];
	abilities: IPokemonAbilities[];
	stats: IPokemonStats[];
}
