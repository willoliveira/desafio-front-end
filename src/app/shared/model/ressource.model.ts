export interface IRessource {
	name: string;
	url: string;
}
