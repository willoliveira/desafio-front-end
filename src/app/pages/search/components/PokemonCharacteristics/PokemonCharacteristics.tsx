import React, { FC } from 'react';
import { View, StyleSheet, Text } from 'react-native';
import { IPokemon, IPokemonAbilities } from '../../../../shared/model/pokemon.model';

export interface IPokemonCharacteristicsProps {
	pokemon: IPokemon
}

export const PokemonCharacteristics: FC<IPokemonCharacteristicsProps> = ({ pokemon }: IPokemonCharacteristicsProps) => (
	<View style={styles.characteristicsContainer}>
		<View style={styles.columnContainer}>
			<View style={styles.textContainer}>
				<Text style={styles.nameCharacteristics}>Height</Text>
				<Text style={styles.valueCharacteristics}>{pokemon.height}</Text>
			</View>
			<View style={styles.textContainer}>
				<Text style={styles.nameCharacteristics}>Weight</Text>
				<Text style={styles.valueCharacteristics}>{pokemon.weight}</Text>
			</View>
		</View>
		<View style={styles.columnContainer}>
			<View style={styles.textContainer}>
				<Text style={styles.nameCharacteristics}>Category</Text>
				<Text style={styles.valueCharacteristics}>-</Text>
			</View>
			<View style={styles.textContainer}>
				<Text style={styles.nameCharacteristics}>Abilities</Text>
				<Text style={styles.valueCharacteristics}>{getAbilities(pokemon.abilities)}</Text>
			</View>
		</View>
	</View>
);

export const getAbilities = (abilities: IPokemonAbilities[]): string => (
	abilities.reduce((acc, ability, index) => {
		if (index === 0) {
			return `${acc}${ability.ability.name}`;
		}
		else if (index + 1 === abilities.length) {
			return `${acc} e ${ability.ability.name}`;
		}
		return `${acc}, ${ability.ability.name}`;
	}, '')
)

const styles = StyleSheet.create({
	characteristicsContainer: {
		flex: 1,
		flexDirection: 'row',
		padding: 12,
		marginBottom: 12,
		backgroundColor: '#30a7d7',
		borderRadius: 8,
		justifyContent: 'space-between'
	},
	columnContainer: {
		flex: 1,
		flexDirection: 'column',
		justifyContent: 'space-between'
	},
	textContainer: {

	},
	nameCharacteristics: {
		color: '#FFFFFF',
		fontWeight: 'bold'
	},
	valueCharacteristics: {
		color: '#212121'
	}
})
