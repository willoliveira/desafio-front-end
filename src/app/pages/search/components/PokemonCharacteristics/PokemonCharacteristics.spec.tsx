/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow } from 'enzyme';
import { PokemonCharacteristics, getAbilities } from './PokemonCharacteristics';
import { IPokemon, IPokemonAbilities } from '../../../../shared/model/pokemon.model';

describe('<PokemonCharacteristics />', () => {
	let pokemon: IPokemon;
	beforeAll(() => {
		pokemon = {
			"abilities": [
				{
					"ability": {
						"name": "chlorophyll",
						"url": "https://pokeapi.co/api/v2/ability/34/"
					},
					"is_hidden": true,
					"slot": 3
				},
				{
					"ability": {
						"name": "overgrow",
						"url": "https://pokeapi.co/api/v2/ability/65/"
					},
					"is_hidden": false,
					"slot": 1
				}
			],
			"base_experience": 64,
			"height": 7,
			"id": 1,
			"name": "bulbasaur",
			"sprites": {
				"back_default": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/1.png",
				"back_female": null,
				"back_shiny": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/back/shiny/1.png",
				"back_shiny_female": null,
				"front_default": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
				"front_female": null,
				"front_shiny": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/shiny/1.png",
				"front_shiny_female": null
			},
			"stats": [
				{
					"base_stat": 45,
					"effort": 0,
					"stat": {
						"name": "speed",
						"url": "https://pokeapi.co/api/v2/stat/6/"
					}
				},
				{
					"base_stat": 65,
					"effort": 0,
					"stat": {
						"name": "special-defense",
						"url": "https://pokeapi.co/api/v2/stat/5/"
					}
				},
				{
					"base_stat": 65,
					"effort": 1,
					"stat": {
						"name": "special-attack",
						"url": "https://pokeapi.co/api/v2/stat/4/"
					}
				},
				{
					"base_stat": 49,
					"effort": 0,
					"stat": {
						"name": "defense",
						"url": "https://pokeapi.co/api/v2/stat/3/"
					}
				},
				{
					"base_stat": 49,
					"effort": 0,
					"stat": {
						"name": "attack",
						"url": "https://pokeapi.co/api/v2/stat/2/"
					}
				},
				{
					"base_stat": 45,
					"effort": 0,
					"stat": {
						"name": "hp",
						"url": "https://pokeapi.co/api/v2/stat/1/"
					}
				}
			],
			"types": [
				{
					"slot": 2,
					"type": {
						"name": "poison",
						"url": "https://pokeapi.co/api/v2/type/4/"
					}
				},
				{
					"slot": 1,
					"type": {
						"name": "grass",
						"url": "https://pokeapi.co/api/v2/type/12/"
					}
				}
			],
			"weight": 69
		}
	})

	describe('Rendering', () => {

		it('should match to snapshot', () => {
			const component = shallow(<PokemonCharacteristics pokemon={pokemon} />);
			expect(component).toMatchSnapshot("PokemonType snapshot");
		});
	});

	describe('getAbilities()', () => {

		it('one abilities', () => {
			const abilities: IPokemonAbilities[] = [
				{
					"ability": {
						"name": "synchronize",
						"url": "https://pokeapi.co/api/v2/ability/28/"
					},
					"is_hidden": false,
					"slot": 1
				}
			];
			expect(getAbilities(abilities)).toEqual('synchronize');
		});

		it('two abilities', () => {
			expect(getAbilities(pokemon.abilities)).toEqual('chlorophyll e overgrow');
		});

		it('three abilities', () => {
			const abilities: IPokemonAbilities[] = [
				{
					"ability": {
						"name": "defiant",
						"url": "https://pokeapi.co/api/v2/ability/128/"
					},
					"is_hidden": true,
					"slot": 3
				},
				{
					"ability": {
						"name": "anger-point",
						"url": "https://pokeapi.co/api/v2/ability/83/"
					},
					"is_hidden": false,
					"slot": 2
				},
				{
					"ability": {
						"name": "vital-spirit",
						"url": "https://pokeapi.co/api/v2/ability/72/"
					},
					"is_hidden": false,
					"slot": 1
				}
			];
			expect(getAbilities(abilities)).toEqual('defiant, anger-point e vital-spirit');
		});
	})
});
