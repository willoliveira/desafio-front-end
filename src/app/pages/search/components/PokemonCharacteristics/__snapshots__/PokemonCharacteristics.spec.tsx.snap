// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonCharacteristics /> Rendering should match to snapshot: PokemonType snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#30a7d7",
      "borderRadius": 8,
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginBottom": 12,
      "padding": 12,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontWeight": "bold",
          }
        }
      >
        Height
      </Text>
      <Text
        style={
          Object {
            "color": "#212121",
          }
        }
      >
        7
      </Text>
    </View>
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontWeight": "bold",
          }
        }
      >
        Weight
      </Text>
      <Text
        style={
          Object {
            "color": "#212121",
          }
        }
      >
        69
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontWeight": "bold",
          }
        }
      >
        Category
      </Text>
      <Text
        style={
          Object {
            "color": "#212121",
          }
        }
      >
        -
      </Text>
    </View>
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontWeight": "bold",
          }
        }
      >
        Abilities
      </Text>
      <Text
        style={
          Object {
            "color": "#212121",
          }
        }
      >
        chlorophyll e overgrow
      </Text>
    </View>
  </View>
</View>
`;
